grpc/shared_conn.go

```diff
@@ -18,68 +24,190 @@ type OnTrackCB func(tr *webrtc.TrackRemote, r *webrtc.RTPReceiver)
-func NewSharedConn(conn *ReconfigurableClientConn, peerConn *webrtc.PeerConnection, logger logging.Logger) *SharedConn {
-	sc := &SharedConn{
-		ReconfigurableClientConn: conn,
-		peerConn:                 peerConn,
-		resourceOnTrackCBs:       map[resource.Name]OnTrackCB{},
-		logger:                   logger,
-	}
-	if peerConn != nil {
-		peerConn.OnTrack(func(tr *webrtc.TrackRemote, r *webrtc.RTPReceiver) {
-			name, err := resource.NewFromString(tr.StreamID())
-			if err != nil {
-				logger.Errorf("%p OnTrack called with StreamID: %s which is not able to be parsed into a resource name", sc, tr.StreamID())
-				return
-			}
-			sc.mu.RLock()
-			onTrackCB, ok := sc.resourceOnTrackCBs[name]
-			sc.mu.RUnlock()
-			if !ok {
-				sc.logger.Errorf("%p OnTrack called with StreamID: %s which is not in the resourceOnTrackCBs: %#v", sc, name, sc.resourceOnTrackCBs)
-				return
-			}
-			onTrackCB(tr, r)
-		})
-	}
-	return sc
+// Invoke forwards to the underlying GRPC Connection.
+func (sc *SharedConn) Invoke(
+	ctx context.Context,
+	method string,
+	args, reply interface{},
+	opts ...googlegrpc.CallOption,
+) error {
+	return sc.grpcConn.Invoke(ctx, method, args, reply, opts...)
+}
+
+// NewStream forwards to the underlying GRPC Connection.
+func (sc *SharedConn) NewStream(
+	ctx context.Context,
+	desc *googlegrpc.StreamDesc,
+	method string,
+	opts ...googlegrpc.CallOption,
+) (googlegrpc.ClientStream, error) {
+	return sc.grpcConn.NewStream(ctx, desc, method, opts...)
 }
 
 // AddOnTrackSub adds an OnTrack subscription for the resource.
 func (sc *SharedConn) AddOnTrackSub(name resource.Name, onTrackCB OnTrackCB) {
-	sc.mu.Lock()
-	defer sc.mu.Unlock()
-	sc.resourceOnTrackCBs[name] = onTrackCB
+	sc.resOnTrackMu.Lock()
+	defer sc.resOnTrackMu.Unlock()
+	sc.resOnTrackCBs[name] = onTrackCB
 }
 
 // RemoveOnTrackSub removes an OnTrack subscription for the resource.
 func (sc *SharedConn) RemoveOnTrackSub(name resource.Name) {
-	sc.mu.Lock()
-	defer sc.mu.Unlock()
-	delete(sc.resourceOnTrackCBs, name)
+	sc.resOnTrackMu.Lock()
+	defer sc.resOnTrackMu.Unlock()
+	delete(sc.resOnTrackCBs, name)
+}
+
+// GrpcConn returns a gRPC capable client connection.
+func (sc *SharedConn) GrpcConn() *ReconfigurableClientConn {
+	return &sc.grpcConn
 }
 
-// PeerConn returns the PeerConnection.
+// PeerConn returns a WebRTC PeerConnection capable of sending video/audio data.
 func (sc *SharedConn) PeerConn() *webrtc.PeerConnection {
 	return sc.peerConn
 }
 
+// ResetConn replaces the underlying GrpcConnection object. It also re-initiatlizes the
+// PeerConnection that must be renegotiated.
+func (sc *SharedConn) ResetConn(conn rpc.ClientConn, moduleLogger logging.Logger) {
+	sc.grpcConn.ReplaceConn(conn)
+
+	sc.logger = moduleLogger.Sublogger("conn")
+
+	peerConn, err := webrtc.NewPeerConnection(webrtc.Configuration{})
+	if err != nil {
+		sc.logger.Warnw("Unable to create optional peer connection for module. Ignoring.", "err", err)
+		return
```

Comment:
If there is already a peer connection on `sc`, won't it be left in a zombie state where it `sc` has a new grpc conn but a stale webrtc conn?

Should we maybe clear & close the webrtc conn whenever ResetConn starts?

Reply:
I think nilling out makes sense.

I have some personal FUD where unconditionally closing the existing connections forces us to cope with some relaxed states that would mask programming bugs. E.g: if this side of the connection is just replacing stuff without it being closed first, why would we expect the other end to be in a state that's looking to accept a new connection?

I also believe having code paths that never get exercised in practice is a maintainability problem. Though defensive programming has its utility as well.

I do think there's some room in being able to use nil-ness to detect if we're going through an unexpected state change. I'm also concluding we should serialize access to these connection methods. So making that addition now as well.